### PR TITLE
Merged branch

### PR DESCRIPTION
**Pull request by ignaciotb**
Mainly functions in benchmark.hpp overloaded to work with PointsT as input instead of mbes_pings.
I've broken down some of the previous implementations to reuse code.
Also, I've added a couple of functions to navi_data.hpp that I've been using

### CODE REVIEW

File: src/data_tools/src/benchmark.cpp [hunk 1]
```diff
@@ -498,8 +664,8 @@ mbes_ping::PingsT registration_summary_benchmark::get_submap_pings_index(const m
     return pings_i;
 }
 
-pair<double, cv::Mat> track_error_benchmark::compute_draw_error_consistency_map(mbes_ping::PingsT& pings)
-{
+vector<vector<vector<Eigen::MatrixXd>>> track_error_benchmark::create_grids_from_pings(mbes_ping::PingsT& pings){
```
Comment (nilsbore) [inline src/data_tools/src/benchmark.cpp]: ">>>" will not work with some c++ compilers, there needs to be spaces in between.

File: src/data_tools/src/navi_data.cpp [hunk 2]
```diff
@@ -66,9 +66,134 @@ void match_timestamps(mbes_ping::PingsT& pings, nav_entry::EntriesT& entries)
             //cout << "Ping " << ping.time_string_ << " closest to Entry " << pos->time_string_ << endl;
         }
     }
+}
+
+
+void saveSubmapsFiles(const std_data::mbes_ping::PingsT& pings, const boost::filesystem::path& folder){
```
Comment (nilsbore) [inline src/data_tools/src/navi_data.cpp]: This function is not using the same naming scheme as the rest

File: src/data_tools/src/navi_data.cpp [hunk 3]
```diff
@@ -66,9 +66,134 @@ void match_timestamps(mbes_ping::PingsT& pings, nav_entry::EntriesT& entries)
...
+        if(myfile.is_open()){
+            for(unsigned int i=0; i<ping_i.beams.size(); i++){
+                myfile << ping_i.beams.at(i)(0) - 669000.0 << " " << ping_i.beams.at(i)(1) - 6383000.0 << " " << ping_i.beams.at(i)(2) << "\n";
+            }
+        }
+    }
+}
+
+
+double computeInfoInSubmap(std_data::mbes_ping::PingsT& submap_pings){
```
Comment (nilsbore) [inline src/data_tools/src/navi_data.cpp]: Not this one either.

File: src/data_tools/src/navi_data.cpp [hunk 4]
```diff
@@ -66,9 +66,134 @@ void match_timestamps(mbes_ping::PingsT& pings, nav_entry::EntriesT& entries)
...
+    double cond_num = 0;
+    for (const std_data::mbes_ping& ping: submap_pings){
+        for(const Eigen::Vector3d& beam_i: ping.beams){
+            cond_num += std::abs(beam_i[2] - mean_beam);
+        }
+    }
+    return cond_num = cond_num / beam_cnt;
+}
+
+void divide_tracks_adaptively(mbes_ping::PingsT& pings)
```
Comment (nilsbore) [inline src/data_tools/src/navi_data.cpp]: Please change to "adaptive"

File: src/data_tools/include/data_tools/benchmark.h [hunk 5]
```diff
@@ -41,6 +43,8 @@ struct track_error_benchmark {
     std::map<std::string, double> consistency_rms_errors;
     double min_consistency_error;
     double max_consistency_error;
+    double min_depth_;
+    double max_depth_;
```
Comment (nilsbore) [inline src/data_tools/include/data_tools/benchmark.h]: Why are these needed in the struct?

File: src/data_tools/include/data_tools/benchmark.h [hunk 6]
```diff
@@ -70,13 +74,29 @@ struct track_error_benchmark {
...
+    void add_ground_truth(PointsT &map_points, PointsT &track_points);
+    void add_benchmark(PointsT &maps_points, PointsT &tracks_points, const std::string &name);
+
+    void track_img_params(PointsT& points_maps, int rows=1000, int cols=1000);
+    cv::Mat draw_height_map(PointsT &points_maps);
+    std::vector<std::vector<std::vector<Eigen::MatrixXd>>> create_grids_from_pings(std_data::mbes_ping::PingsT& pings);
+    std::vector<std::vector<std::vector<Eigen::MatrixXd>>> create_grids_from_matrices(PointsT& points_maps);
+    std::pair<double, Eigen::MatrixXd> compute_consistency_error(
+            std::vector<std::vector<std::vector<Eigen::MatrixXd>>>& grid_maps);
+    cv::Mat draw_error_consistency_map(Eigen::MatrixXd values);
```
Comment (nilsbore) [inline src/data_tools/include/data_tools/benchmark.h]: I think we still need the "compute_draw_error_consistency_map" function for all of our old code to still work? Did you remove that?

File: src/data_tools/include/data_tools/benchmark.h [hunk 7]
```diff
@@ -70,13 +74,29 @@ struct track_error_benchmark {
     void add_benchmark(std_data::pt_submaps::TransT& trans_corr, std_data::pt_submaps::RotsT& rots_corr, const std::string& name);
     void print_summary();
 
+    // Overloaded functions to work with input submaps in PointsT format
+    void add_ground_truth(PointsT &map_points, PointsT &track_points);
+    void add_benchmark(PointsT &maps_points, PointsT &tracks_points, const std::string &name);
+
+    void track_img_params(PointsT& points_maps, int rows=1000, int cols=1000);
+    cv::Mat draw_height_map(PointsT &points_maps);
+    std::vector<std::vector<std::vector<Eigen::MatrixXd>>> create_grids_from_pings(std_data::mbes_ping::PingsT& pings);
```
Comment (nilsbore) [inline src/data_tools/include/data_tools/benchmark.h]: Same thing with ">>>" for these two functions.

File: src/data_tools/src/benchmark.cpp [hunk 8]
```diff
@@ -263,6 +248,92 @@ pair<double, cv::Mat> track_error_benchmark::compute_draw_consistency_map(mbes_p
     return make_pair(meanv, error_img);
 }
 
+void track_error_benchmark::map_draw_params(PointsT& map_points, PointsT& track_points,
```
Comment (nilsbore) [inline src/data_tools/src/benchmark.cpp]: What does this function do?

File: src/data_tools/src/benchmark.cpp [hunk 9]
```diff
@@ -538,6 +706,54 @@ pair<double, cv::Mat> track_error_benchmark::compute_draw_error_consistency_map(
         pos = next;
     }
 
+    return grid_maps;
+}
+
+vector<vector<vector<Eigen::MatrixXd>>> track_error_benchmark::create_grids_from_matrices(PointsT& points_maps){
```
Comment (nilsbore) [inline src/data_tools/src/benchmark.cpp]: ">>>"

File: src/data_tools/src/benchmark.cpp [hunk 10]
```diff
@@ -590,6 +809,7 @@ pair<double, cv::Mat> track_error_benchmark::compute_draw_error_consistency_map(
                         continue;
                     }
                     Eigen::Vector3d point = grid_maps[i][j][m].row(rand()%grid_maps[i][j][m].rows()).transpose();
+//                    Eigen::Vector3d point = grid_maps[i][j][m].row(((int)grid_maps[i][j][m].rows()/10)*c).transpose();
```
Comment (nilsbore) [inline src/data_tools/src/benchmark.cpp]: Why is this comment here?

File: src/data_tools/src/benchmark.cpp [hunk 11]
```diff
@@ -538,6 +706,54 @@ pair<double, cv::Mat> track_error_benchmark::compute_draw_error_consistency_map(
...
+                grid_maps[row][col][k].conservativeResize(grid_maps[row][col][k].rows()+1, 3);
+                grid_maps[row][col][k].bottomRows<1>() = point_i.transpose();
+            }
+        }
+        k++;
+    }
+    return grid_maps;
+}
+
+std::pair<double, Eigen::MatrixXd> track_error_benchmark::compute_consistency_error(
```
Comment (nilsbore) [inline src/data_tools/src/benchmark.cpp]: ">>>"

File: src/data_tools/src/benchmark.cpp [hunk 12]
```diff
@@ -321,20 +444,63 @@ void track_error_benchmark::add_ground_truth(mbes_ping::PingsT& pings)
...
 {
     input_pings = pings;
 }
 
+void track_error_benchmark::add_benchmark(PointsT& maps_points, PointsT& tracks_points,
+                                          const std::string& name){
+    cv::Mat error_img;
+    Eigen::MatrixXd error_vals;
+    double consistency_rms_error;
+    std::vector<std::vector<std::vector<Eigen::MatrixXd>>> grid_maps = create_grids_from_matrices(maps_points);
```
Comment (nilsbore) [inline src/data_tools/src/benchmark.cpp]: I like how you did this to enable using the same consistency thing for both. Again, ">>>".

File: src/data_tools/src/benchmark.cpp [hunk 13]
```diff
@@ -321,20 +444,63 @@ void track_error_benchmark::add_ground_truth(mbes_ping::PingsT& pings)
...
     int nbr_tracks_drawn = track_rms_errors.size();
     cv::Scalar color(colormap[nbr_tracks_drawn][2], colormap[nbr_tracks_drawn][1], colormap[nbr_tracks_drawn][0]);
     draw_track_img(pings, track_img, color, name);
     cv::Mat error_img;
+    Eigen::MatrixXd error_vals;
     double consistency_rms_error;
-    //tie(consistency_rms_error, error_img) = compute_draw_consistency_map(pings);
-    tie(consistency_rms_error, error_img) = compute_draw_error_consistency_map(pings);
+    tie(consistency_rms_error, error_img) = compute_draw_consistency_map(pings);
+    std::vector<std::vector<std::vector<Eigen::MatrixXd>>> grid_maps = create_grids_from_pings(pings);
```
Comment (nilsbore) [inline src/data_tools/src/benchmark.cpp]: ">>>"

File: src/data_tools/include/data_tools/benchmark.h (hunk 6 above)
Comment (ignaciotb) [reply src/data_tools/include/data_tools/benchmark.h]: I split it up into two functions (to reuse some code). afaik that function was only used within add_benchmark(). If so, it isn't needed it anymore.

**Comment by nilsbore** [comment]
How do you see in what way you resolved the comments? This interface is a bit weird...

**Comment by ignaciotb** [comment]
image
Click View changes here

**Comment by nilsbore** [comment]
You had to look at the diff, and then filter by changes since your last review.

File: src/data_tools/src/benchmark.cpp (hunk 8 above)
Comment (ignaciotb) [reply src/data_tools/src/benchmark.cpp]: It's equivalent to track_img_params() but it saves the parameters from the whole map to be used when drawing the heightmap of each submap independently
